src/build.cc

```diff
@@ -679,14 +687,25 @@ bool Builder::StartEdge(Edge* edge, string* err) {
 
   status_->BuildEdgeStarted(edge, start_time_millis);
 
-  // Create directories necessary for outputs.
+  TimeStamp build_start = -1;
+
+  // Create directories necessary for outputs and remember the current
+  // filesystem mtime to record later
   // XXX: this will block; do we care?
   for (vector<Node*>::iterator o = edge->outputs_.begin();
        o != edge->outputs_.end(); ++o) {
     if (!disk_interface_->MakeDirs((*o)->path()))
       return false;
+    if (build_start == -1) {
+      disk_interface_->WriteFile(tmp_mtime_file_path_, "");
```

Comment:
Which command updates the mtime in `RealDiskInterface::WriteFile`? Couldn't `fwrite` just do nothing if it was passed an empty buffer?

Reply:
fwrite will update the file to be empty if it is passed an empty buffer, it can't simply decide to do nothing.

Reply:
what if it is already empty?

Reply:
It will still write the empty string to the file and update the mtime.

Reply:
Is this documented specified somewhere?

Wouldn't matter if we had real-world tests we could run on all systems :(

Reply:
https://en.cppreference.com/w/c/io/fwrite

While it's ultimately up to the filesystem implementation to update the mtime if a file is written to, fwrite() does guarantee that the file is written to. We specifically take advantage of using the filesystem for acquiring an mtime because we compare this acquired time to the input mtimes on subsequent runs, which are presumably residing on the same filesystem (or one with times that are closely in sync).

Reply:
> If size or count is zero, fwrite returns zero and performs no other action.

Sounds like it doesn't write to the file.

Reply:
Apologies, it's not `fwrite()` that causes the mtime to be updated, it's the `fopen(path, "w")` call that does so. The fact that we're writing 0 bytes is actually an optimization that means the `fwrite()` call itself is a no-op (like you say).

`fopen()` creates the file if it does not exist, or deletes (overwrites) the file if it already exists. Either way, mtime is updated.